### PR TITLE
Add shopt decorator

### PR DESCRIPTION
**Pull request by antevens**
I needed to get rid of mapfile to avoid errors in bash v3 for Ellen

### CODE REVIEW

File: shtdlib.sh [hunk 1]
```diff
@@ -178,6 +179,56 @@ function umask_decorator {
...
 
+# Bash behaviour option decorator
+# Allows changing/setting bash options for a function code block restoring
+# the original once execution of the function and it's calls is complete.
+# Requires an option name (see shopt) and a truthyness vale "true"/"false" or
+# other command that returns 0/1.
+# variables shopt_decorator_option_name and shopt_decorator_option_value
+# To use this add a line like the following (without #) as the first line of a function
+# shopt_decorator "${FUNCNAME[0]}" "${@:-}" && return
+function shopt_decorator {
```
Comment (mroote) [inline shtdlib.sh]: Should the `decorator_option_value` and `decorator_option_name` variables be passed in as positional args instead?

Comment (mroote) [reply shtdlib.sh]: Ah ok that makes sense.

File: shtdlib.sh [hunk 2]
```diff
@@ -178,6 +179,56 @@ function umask_decorator {
...
 }
 
+# Bash behaviour option decorator
+# Allows changing/setting bash options for a function code block restoring
+# the original once execution of the function and it's calls is complete.
+# Requires an option name (see shopt) and a truthyness vale "true"/"false" or
+# other command that returns 0/1.
+# variables shopt_decorator_option_name and shopt_decorator_option_value
+# To use this add a line like the following (without #) as the first line of a function
+# shopt_decorator "${FUNCNAME[0]}" "${@:-}" && return
```
Comment (hrpatel) [inline shtdlib.sh]: We should provide a full example. 

```bash
function smarter_sort {
    # 'sort' doesn't properly handle SIGPIPE
    shopt_decorator_option_name='pipefail'
    shopt_decorator_option_value='false'
    shopt_decorator "${FUNCNAME[0]}" "${@:-}" && return

    # Do some sorting
}
```

File: shtdlib.sh [hunk 3]
```diff
@@ -178,6 +179,56 @@ function umask_decorator {
...
+function shopt_decorator {
+    if [ -n "${shopt_decorator_option_value:-}" ]  && [ -n "$(shopt -o "${shopt_decorator_option_name:-}")" ] ; then
+        if [ "${FUNCNAME[0]}" != "${FUNCNAME[2]}" ] ; then
+            if shopt -qo "${shopt_decorator_option_name}" ; then
```
Comment (hrpatel) [inline shtdlib.sh]: I completely missed the `-q` flag - that would have simplified my code so much!

File: shtdlib.sh [hunk 4]
```diff
@@ -178,6 +179,56 @@ function umask_decorator {
...
+                shopt -so "${shopt_decorator_option_name}"
+                return 0
+            else
+                # Option is not set
+                if "${shopt_decorator_option_value}" ; then
+                    # Option should be set
+                    debug 10 "Temporarily setting bash option ${shopt_decorator_option_name}"
+                    shopt -so "${shopt_decorator_option_name}"
+                else
+                    debug 10 "No need to set ${shopt_decorator_option_name}, it's already ${shopt_decorator_option_value}"
```
Comment (hrpatel) [inline shtdlib.sh]: `s/set/unset/`

File: shtdlib.sh [hunk 5]
```diff
@@ -564,6 +530,10 @@ function exit_on_fail {
 # add_on_mod callback "${path_to_monitor}"
 #
 function add_on_mod {
+shopt_decorator_option_name='nounset'
```
Comment (hrpatel) [inline shtdlib.sh]: Indent required

File: shtdlib.sh [hunk 6]
```diff
@@ -178,6 +179,56 @@ function umask_decorator {
     return 1
 }
 
+# Bash behaviour option decorator
+# Allows changing/setting bash options for a function code block restoring
+# the original once execution of the function and it's calls is complete.
+# Requires an option name (see shopt) and a truthyness vale "true"/"false" or
```
[the remark this answers is not in the document]
Comment (antevens) [reply shtdlib.sh]: It's like a valley only more picturesque.

File: shtdlib.sh (hunk 3 above)
Comment (antevens) [reply shtdlib.sh]: I know, I saw that and thought the same.

File: shtdlib.sh (hunk 2 above)
Comment (antevens) [reply shtdlib.sh]: Thanks for the example!

File: shtdlib.sh (hunk 4 above)
Comment (antevens) [reply shtdlib.sh]: nope, it's supposed to be set, not unset (shopt -so sets a value, shopt -uo unsets)

**Pull request merged**